// @sphinx-labs/core/dist/config/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionInputType = exports.ContractKindEnum = exports.contractKindHashes = exports.userContractKinds = void 0;
const contracts_1 = require("@sphinx-labs/contracts");
exports.userContractKinds = [
    'oz-transparent',
    'oz-ownable-uups',
    'oz-access-control-uups',
    'external-transparent',
    'immutable',
    'proxy',
];
exports.contractKindHashes = {
    'external-transparent': contracts_1.EXTERNAL_TRANSPARENT_PROXY_TYPE_HASH,
    'oz-transparent': contracts_1.OZ_TRANSPARENT_PROXY_TYPE_HASH,
    'oz-ownable-uups': contracts_1.OZ_UUPS_OWNABLE_PROXY_TYPE_HASH,
    'oz-access-control-uups': contracts_1.OZ_UUPS_ACCESS_CONTROL_PROXY_TYPE_HASH,
    immutable: contracts_1.IMMUTABLE_TYPE_HASH,
    implementation: contracts_1.IMPLEMENTATION_TYPE_HASH,
    proxy: contracts_1.DEFAULT_PROXY_TYPE_HASH,
};
var ContractKindEnum;
(function (ContractKindEnum) {
    ContractKindEnum[ContractKindEnum["INTERNAL_DEFAULT"] = 0] = "INTERNAL_DEFAULT";
    ContractKindEnum[ContractKindEnum["OZ_TRANSPARENT"] = 1] = "OZ_TRANSPARENT";
    ContractKindEnum[ContractKindEnum["OZ_OWNABLE_UUPS"] = 2] = "OZ_OWNABLE_UUPS";
    ContractKindEnum[ContractKindEnum["OZ_ACCESS_CONTROL_UUPS"] = 3] = "OZ_ACCESS_CONTROL_UUPS";
    ContractKindEnum[ContractKindEnum["EXTERNAL_DEFAULT"] = 4] = "EXTERNAL_DEFAULT";
    ContractKindEnum[ContractKindEnum["IMMUTABLE"] = 5] = "IMMUTABLE";
})(ContractKindEnum = exports.ContractKindEnum || (exports.ContractKindEnum = {}));
var ActionInputType;
(function (ActionInputType) {
    ActionInputType[ActionInputType["CREATE"] = 0] = "CREATE";
    ActionInputType[ActionInputType["CREATE2"] = 1] = "CREATE2";
    ActionInputType[ActionInputType["CALL"] = 2] = "CALL";
})(ActionInputType = exports.ActionInputType || (exports.ActionInputType = {}));
