// @sphinx-labs/contracts/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RECOMMENDED_REMAPPING = exports.EXECUTION_LOCK_TIME = exports.CREATE3_PROXY_INITCODE = exports.DETERMINISTIC_DEPLOYMENT_PROXY_ADDRESS = exports.DEFAULT_PROXY_TYPE_HASH = exports.IMPLEMENTATION_TYPE_HASH = exports.IMMUTABLE_TYPE_HASH = exports.OZ_UUPS_ACCESS_CONTROL_PROXY_TYPE_HASH = exports.OZ_UUPS_OWNABLE_PROXY_TYPE_HASH = exports.OZ_TRANSPARENT_PROXY_TYPE_HASH = exports.EXTERNAL_TRANSPARENT_PROXY_TYPE_HASH = exports.SPHINX_PROXY_ADMIN_ADDRESS_SLOT_KEY = exports.DEFAULT_ADMIN_ROLE = exports.EXECUTOR = exports.getOwnerAddress = exports.OWNER_MULTISIG_ADDRESS = void 0;
const ethers_1 = require("ethers");
exports.OWNER_MULTISIG_ADDRESS = '0x226F14C3e19788934Ff37C653Cf5e24caD198341';
const getOwnerAddress = () => {
    return process.env.SPHINX_INTERNAL__OWNER_PRIVATE_KEY
        ? new ethers_1.ethers.Wallet(process.env.SPHINX_INTERNAL__OWNER_PRIVATE_KEY).address
        : exports.OWNER_MULTISIG_ADDRESS;
};
exports.getOwnerAddress = getOwnerAddress;
exports.EXECUTOR = '0x42761facf5e6091fca0e38f450adfb1e22bd8c3c';
exports.DEFAULT_ADMIN_ROLE = ethers_1.ethers.ZeroHash;
exports.SPHINX_PROXY_ADMIN_ADDRESS_SLOT_KEY = (0, ethers_1.toBeHex)(BigInt(ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('sphinx.proxy.admin'))) - BigInt(1));
exports.EXTERNAL_TRANSPARENT_PROXY_TYPE_HASH = ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('external-transparent'));
exports.OZ_TRANSPARENT_PROXY_TYPE_HASH = ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('oz-transparent'));
exports.OZ_UUPS_OWNABLE_PROXY_TYPE_HASH = ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('oz-ownable-uups'));
exports.OZ_UUPS_ACCESS_CONTROL_PROXY_TYPE_HASH = ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('oz-access-control-uups'));
exports.IMMUTABLE_TYPE_HASH = ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('immutable'));
exports.IMPLEMENTATION_TYPE_HASH = ethers_1.ethers.keccak256(ethers_1.ethers.toUtf8Bytes('implementation'));
exports.DEFAULT_PROXY_TYPE_HASH = ethers_1.ethers.ZeroHash;
exports.DETERMINISTIC_DEPLOYMENT_PROXY_ADDRESS = '0x4e59b44847b379578588920cA78FbF26c0B4956C';
exports.CREATE3_PROXY_INITCODE = '0x67363d3d37363d34f03d5260086018f3';
exports.EXECUTION_LOCK_TIME = 15 * 60;
exports.RECOMMENDED_REMAPPING = '@sphinx-labs/contracts/=lib/sphinx/packages/contracts/contracts/foundry';
